KEYCLOAK-18466 Configure HTTP client timeouts for adapters

mabartos opened the pull request:
JIRA: [KEYCLOAK-18466](https://issues.redhat.com/browse/KEYCLOAK-18466)

IMHO, it should be possible to customize timeouts for HTTP client included in the adapter core because each application/servlet requires various requirements. You can define these timeouts in XML (SAML) and JSON (OIDC) adapter configuration file. From QE perspective, some tests executed on particular machines require change of these timeouts in order to run properly.

@hmlnarik @mhajas Could you please take a look at it?

ruromero committed:
[KEYCLOAK-17872] Add missing HTTPClient properties

Signed-off-by: ruromero <rromerom@redhat.com>

mabartos committed:
KEYCLOAK-18466 Configure HTTP client timeouts for adapters

mhajas commented on adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:
```diff
@@ -0,0 +1,550 @@
+                <xs:documentation>Password for the truststore keystore.</xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="socketTimeout" type="xs:long" default="-1">
```
Do we need `long` here? 9223372036854775807 sounds like an overkill for the socket timeout. Maybe, we should follow the [API of httpclient](https://hc.apache.org/httpclient-legacy/apidocs/org/apache/commons/httpclient/params/HttpConnectionParams.html#setSoTimeout(int)) and use int?

mabartos replied:
I agree with you. I just wanted to keep the same type of the timeout variable and keep it consistent with our codebase. But it makes sense to use only integer type for this.

mabartos replied:
I agree with you. I just wanted to keep the same type of the timeout variable and keep it consistent with our codebase. But it makes sense to use only integer type for this.

mhajas replied:
@mabartos I had a look at the long vs int again and I think we should revert to long, as we are using it in server code as well. The reason probably is, that it is also possible to configure TimeUnit which then decreases the maximum value. Also as you said, we should stay consistent. Sorry for the confusion.

mhajas commented on adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:
```diff
@@ -376,6 +376,14 @@ public HttpClient build(AdapterHttpClientConfig adapterConfig) {
 
         configureProxyForAuthServerIfProvided(adapterConfig);
 
+        if (socketTimeout == -1 && adapterConfig.getSocketTimeout() != -1) {
```
```suggestion
        if (socketTimeout == -1 && adapterConfig.getSocketTimeout() > 0) {
```

mabartos replied:
Probably it's better condition. I assume, nobody wants to set socket timeout to 0; therefore in this case, the default value should be used.

mhajas replied:
It seems that zero is the default value which corresponds to infinite timeout.

mhajas commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java:
```diff
@@ -210,99 +212,104 @@
     @Rule
     public AssertEvents assertEvents = new AssertEvents(this);
 
+    private static final Consumer<AdapterConfig> GLOBAL_ADAPTER_CONFIG = config -> {
```
Is there some instability in this test? I don't see it from the Jira? Should we link another Jira to the PR?

mabartos replied:
I've found out the instabilities, which this PR should solve are related to whole platform and there's created another issue (see #8175 ). Thus, this is out of date and I'll remove it.

mhajas reviewed (commented):
Thank you for the PR @mabartos. Overall, the changes looks good to me, however I added few comments in the code. 

However, this is probably something we should discuss on mailing list, maybe system variable way of configuring this would be preferred as you suggested here: https://github.com/keycloak/keycloak/pull/8175.

mabartos commented on adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:
```diff
@@ -0,0 +1,550 @@
+            <xs:element name="RoleIdentifiers" type="role-identifiers-type" minOccurs="0" maxOccurs="1">
+                <xs:annotation>
+                    <xs:documentation>Defines what SAML attributes within the assertion received from the user should be used as role identifiers within the Java EE Security Context for the user.
+                        By default Role attribute values are converted to Java EE roles. Some IDPs send roles via a member or memberOf attribute assertion. You can define one or more Attribute elements to specify which SAML attributes must be converted into roles.</xs:documentation>
```
Ok.

mabartos commented:
@mhajas Thanks for your comment. I'll fix this.

I've just started thread on Keycloak dev (See [here](https://groups.google.com/g/keycloak-dev/c/HLLo9wk0dSI)).

hmlnarik commented on adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:
```diff
@@ -113,9 +113,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected TimeUnit connectionTTLUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Why this change?

mabartos replied:
I've been discussing it with @mhajas and considering to change from _long_ to _int_ in order to follow Apache HTTP client API. However, as you've mentioned, this is unrelated to the JIRA and the change is not necessary though.

hmlnarik replied:
Please keep this PR minimal, it is already big enough.

hmlnarik commented on adapters/saml/core/src/main/java/org/keycloak/adapters/cloned/HttpClientBuilder.java:
```diff
@@ -112,9 +112,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected TimeUnit connectionTTLUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Same here - why this change?

mabartos replied:
Same as above.

hmlnarik commented on adapters/saml/core/src/test/java/org/keycloak/adapters/saml/config/parsers/KeycloakSamlAdapterXMLParserTest.java:
```diff
@@ -117,57 +117,59 @@ public void testParseSimpleFileNoNamespace() throws Exception {
+        assertThat(sp.isIsPassive(), is(true));
+        assertThat(sp.isAutodetectBearerOnly(), is(false));
+        assertThat(sp.isKeepDOMAssertion(), is(false));
+        assertThat(sp.getKeys().size(), is(2));
```
Nit:
```suggestion
        assertThat(sp.getKeys(), hasSize(2));
```

mabartos replied:
+1

hmlnarik commented on services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:
```diff
@@ -134,8 +134,8 @@ private void lazyInit(KeycloakSession session) {
                 if (httpClient == null) {
-                    long socketTimeout = config.getLong("socket-timeout-millis", 5000L);
-                    long establishConnectionTimeout = config.getLong("establish-connection-timeout-millis", -1L);
+                    int socketTimeout = config.getInt("socket-timeout-millis", 5000);
```
Why this change?

mabartos replied:
Same as [this](https://github.com/keycloak/keycloak/pull/8165#discussion_r658516576)

hmlnarik commented on services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java:
```diff
@@ -101,9 +101,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected TimeUnit maxConnectionIdleTimeUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Why this change?

mabartos replied:
Same as [this](https://github.com/keycloak/keycloak/pull/8165#discussion_r658516576).

hmlnarik commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:
```diff
@@ -229,4 +247,55 @@ protected void setAdapterServletTimeOffset(int timeOffset, String servletUri) {
+     * Custom Object mapper
+     */
+    private static ObjectMapper getSystemPropertiesObjectMapper() {
+        ObjectMapper mapper = new ObjectMapper(new SystemPropertiesJsonParserFactory());
```
`ObjectMapper` should be a static constant rather than recreated each time.

mabartos replied:
You're right.

hmlnarik commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:
```diff
@@ -65,15 +73,25 @@ protected static WebArchive servletDeploymentMultiTenant(String name, Class... s
     }
 
-    protected static WebArchive servletDeployment(String name, String adapterConfig, Class... servletClasses) {
+    protected static WebArchive servletDeployment(String name, Consumer<AdapterConfig> adapterConfig, Class... servletClasses) {
```
It seems that `adapterConfig` here is always `null`. Where is this method used from? Why is the `Consumer` parameter introduced?

mabartos replied:
In the first commit of this PR, this is used in _DemoServletAdapterTest_, but the solution for instability of the test is out of scope of this PR. So, it's like a preparation for another related tasks and configure adapter directly from a test. Please, see [here](https://github.com/keycloak/keycloak/pull/8165/commits/d1b77b938f54ef1d1070889eb020a79fdf21b503#diff-a7d3a845d41009122b21855f7fa779bbf324c4e6ba58f4404bf4f58addc89cadR215). @hmlnarik WDYT?

hmlnarik replied:
The latter would be better, thank you

mabartos replied:
Ok

hmlnarik reviewed (commented):
Thank you @mabartos. Could you please have a look at the few questions inline? It seems that some changes (changing type) are unrelated to the JIRA, is there any reason for doing so?
Also note that any new saml adapter xsd has to be published to the web, see https://github.com/keycloak/keycloak-web/tree/master/static/schema

mabartos commented:
Created PR to _keycloak-web_. See https://github.com/keycloak/keycloak-web/pull/172

mabartos commented:
@mhajas @hmlnarik Rebased and fixed all your suggestions. After review, these commits can be squashed. I'll execute tests in pipeline and I'll share the results.

hmlnarik commented:
Does it make sense to also include the changes requested in #7968?

mabartos commented:
@hmlnarik I've been thinking about it and I agree with it. I'll merge it.

hmlnarik commented:
Thanks, @mabartos.

I only realized now that there also needs to be a documentation for the added properties:
- https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/saml/java/general-config/idp_httpclient_subelement.adoc
- https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/oidc/java/java-adapter-config.adoc
Apologies I missed this earlier.

mabartos commented:
@hmlnarik Thanks. 
I merged the PR #7968. Thanks @ruromero for your contribution.

PR to _documentation_: https://github.com/keycloak/keycloak-documentation/pull/1220
PR to _keycloak-web_: https://github.com/keycloak/keycloak-web/pull/172

I'll execute the pipeline tests again. Previously, these tests passed in pipeline, therefore this PR didn't break anything.

mabartos commented:
@hmlnarik The pipeline tests passed. See https://master-jenkins.com/job/universal-test-pipeline-adapters/506/

mhajas reviewed (changes requested):
Thanks for all the changes @mabartos. I added one nitpick comment, however, I believe it makes sense to address it. WDYT?

mabartos commented on adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java:
```diff
@@ -101,5 +105,18 @@ public void loadSecretJwtCredentials() {
+        long connectionTimeout = client.getParams().getIntParameter(CoreConnectionPNames.CONNECTION_TIMEOUT, -2);
+
+        assertThat(socketTimeout, CoreMatchers.is(2000L));
+        assertThat(connectionTimeout, CoreMatchers.is(6000L));
```
@mhajas Yeah, sure. Thanks

mabartos replied:
@mhajas However, the `connectionTTL` is only used for internal purposes, especially to create _HttpClient_ with proper _Client Connection Manager_ and with connection pool size greater than 1. The `connectionTTL` attribute is not included in the HttpParams provided by Apache _http-core_ and it seems it shouldn't be revealed in the same manner like those timeouts. In this case, I don't want to create a new parameter which is out of scope of _http-core_ interface. WDYT?

mhajas replied:
@mabartos I am not sure I am following. Could you elaborate a little bit more, please? What is the difference when the configuration option is used in `HttpParams` and in `ClientConnectionManager`? The `ThreadSafeClientConnManager` class comes from `httpclient` as well, so it is not our internal class. Also, we are using the `connectionTTL` option also for [HttpClientProvider](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java#L141) of the Keycloak server.

Are you suggesting removing the configuration of connectionTTL from this PR completely?

mabartos replied:
@mhajas In the `ThreadSafeClientConnManager` is the field `connectionTTL` with private access modifier and it's not possible to obtain it from the class and where it's only used for internal instantiating of pool entry. The field is completely hidden. Anyway, I'd like to find already existing parameter for `connectionTTL` in order to include it to the `HttpParams`, but I think, we will have to create a new one. 

However as you've suggested, if the support for the `connectionTTL` is present in this PR, there should be verified the proper settings of the property. I'll add a new parameter to `HttpParams` in order to obtain configuration externally. 
Thanks.

mabartos commented:
@mhajas I've added _connectionTTL_ parameter to the parameters of the HTTP client. Pipeline tests passed (no new error appeared). See https://master-jenkins.com/job/universal-test-pipeline-adapters/519/

mabartos commented:
After discussion with @mhajas we've decided to get rid off the custom variables for `HttpParams` and in a follow-up task to update whole test class `HttpClientBuilder` for adapters and provide particular test cases. Removed last commit and rebased with master.

mabartos commented:
@mhajas JFYI the pipeline tests passed. See https://master-jenkins.redhat.com/job/universal-test-pipeline-adapters/534/

mhajas reviewed (dismissed):
LGTM @mabartos, thanks for the changes.

mabartos commented:
@hmlnarik JFYI: along with merging this PR, these PRs should be merged as well:

PR to documentation: [keycloak/keycloak-documentation#1220](https://github.com/keycloak/keycloak-documentation/pull/1220)
PR to keycloak-web: [keycloak/keycloak-web#172](https://github.com/keycloak/keycloak-web/pull/172)

hmlnarik commented:
Thank @mabartos, great job! The changes look good to me apart from one inconsistency in letter case - there is already `proxyUrl` and the same letter case should be followed in the `connectionTtl` property (applies to all three PRs)

sguilhen commented:
@mabartos Good work on this one. I just missed the new subsystem XSD for OIDC as7-eap6. Versions were bumped in the subsystem files but version 1_2 of the XSD is missing in this PR.

mabartos commented:
> @mabartos Good work on this one. I just missed the new subsystem XSD for OIDC as7-eap6. Versions were bumped in the subsystem extension (https://github.com/keycloak/keycloak/pull/8165/files#diff-d9314783e37a970e7660f8ed687e723564fcaaefe4a9b5dc2d85607d1c7b43daR40) but version 1_2 of the XSD is missing in this PR.

@sguilhen Nice catch! Thanks.

mabartos committed:
KEYCLOAK-18466 Configure HTTP client timeouts for adapters - change property names

mabartos commented:
@hmlnarik @sguilhen I addressed your comments and rebased with the most recent master. 

After discussion with @hmlnarik, I changed the property names with suffix `-millis` for OIDC. 
For SAML, where the configuration of the HTTP client is defined in XML file, I didn't add the suffix (in this case the property `socketTimeout` would change to `socketTimeoutMillis`), because in the description of the fields, there are already defined the time units; milliseconds. Therefore, IMHO, it's quite counterproductive to define these units again.

I executed the pipeline tests, so the result will be lately.

sguilhen commented:
@mabartos This renaming of the OIDC properties to include the **-millis** suffix creates a mismatch between the SAML properties that are named differently and the JSON properties in AdapterConfig

This means that when the SAML adapter subsystem reads the model and the adapter later on converts the subsystem config into JSON and tries to load AdapterConfig from the converted JSON we have an exception being thrown because the properties don't match. See https://issues.redhat.com/browse/KEYCLOAK-18302 for an example of an issue that I've recently fixed in which the SAML property and the JSON attribute in AdapterConfig had different names.

In the PR for that Jira you can see that I've added a test that configures the problematic property and then tries to convert it into JSON and load the JSON config into AdapterConfig. Without the correct mapping done here https://github.com/keycloak/keycloak/pull/8244/files#diff-a350e9551196cf925e4145d0a0c4e091f0ca2b3f31d28458ec248af121c44e1cR154 the test fails.

mabartos commented:
@sguilhen Thanks for your comment. However, I'm not sure I understand. I'm not aware that SAML adapter converts adapter subsystem configuration to JSON. I'd say it's valid for OIDC adapters. As I saw in your attached PR, everything is related to OIDC, not SAML. For SAML adapter, HTTP client attributes are parsed in class _HttpClientParser_, which are then used in _HttpClientBuilder_ for building the client.

For OIDC Wildfly and AS7 subsystem, there are defined XML properties with the `-millis` suffix same as for JSON properties in _AdapterConfig_, so there shouldn't be any incompatibilities.

@sguilhen Could you please explain it more?

sguilhen commented:
> @sguilhen Thanks for your comment. However, I'm not sure I understand. I'm not aware that SAML adapter converts adapter subsystem configuration to JSON. I'd say it's valid for OIDC adapters. As I saw in your attached PR, everything is related to OIDC, not SAML. For SAML adapter, HTTP client attributes are parsed in class _HttpClientParser_, which are then used in _HttpClientBuilder_ for building the client.
> 
> For OIDC Wildfly and AS7 subsystem, there are defined XML properties with the `-millis` suffix same as for JSON properties in _AdapterConfig_, so there shouldn't be any incompatibilities.
> 
> @sguilhen Could you please explain it more?
> 
> EDIT:
> I added test case for the WF subsystem to verify the JSON parsing.
> See [here](https://github.com/keycloak/keycloak/pull/8165/commits/685fc210d679e2f3332f83fb69bc40c46a7246d3#diff-45be67b9727261060237cf82a49a35364ab157920850906636d6be11f5943fcbR154).

Hi @mabartos You are right, I've got it the other way around, its the OIDC subsystem that converts the model to JSON, SAML works based on XML all the way. This looks good to me!

mabartos commented:
@hmlnarik @mhajas @sguilhen I executed the pipeline tests again in order to ensure that everything works as expected. The pipeline tests passed. See https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/546/

hmlnarik reviewed (approved):
Approving based on @mhajas and @sguilhen 's reviews
